ADO.NET/OAuthSampleApp/OAuthSampleApp.cs

```diff
@@ -0,0 +1,217 @@
+                    await DoPasswordGrant();
+                }
+
+                SetAccessToken();
+                retryCount++;
+            }
+        }
+    }
+
+    // password grant logs into the IDP using credentials in app.config
+    public static async Task DoPasswordGrant()
+    {
+        Dictionary<string, string> formData = new Dictionary<string, string>
+        {
+            {"grant_type",   "password"},
+            {"client_id",     ConfigurationManager.AppSettings["ClientId"]},
+            {"client_secret", ConfigurationManager.AppSettings["ClientSecret"]},
+            {"username",      ConfigurationManager.AppSettings["User"]},
+            {"password",      ConfigurationManager.AppSettings["Password"]}
+        };
+
+        await GetAndSetTokens(formData);
+    }
+
+    // refresh grant uses the refresh token to get a new access and refresh token
+    public static async Task DoTokenRefresh()
+    {
+        Dictionary<string, string> formData = new Dictionary<string, string>
+        {
+            {"grant_type",   "refresh_token"},
+            {"client_id",     ConfigurationManager.AppSettings["ClientId"]},
+            {"client_secret", ConfigurationManager.AppSettings["ClientSecret"]},
+            {"refresh_token", Environment.GetEnvironmentVariable(OAUTH_REFRESH_TOKEN_VAR_STRING, EnvironmentVariableTarget.User)}
+        };
+
+        await GetAndSetTokens(formData);
+    }
+
+    // uses the token url from the config to make an http request
+    // will always always set new access and refresh tokens, even if currently valid
+    private static async Task GetAndSetTokens(Dictionary<string, string> formData)
+    {
+        HttpClient httpClient = new HttpClient();
+
+        try
+        {
+            var content = new FormUrlEncodedContent(formData);
+
+            string tokenUrl = ConfigurationManager.AppSettings["TokenUrl"];
+            var response = await httpClient.PostAsync(tokenUrl, content);
+
+            response.EnsureSuccessStatusCode();
+
+            var result = await response.Content.ReadAsStringAsync();
+            JObject jObject = JObject.Parse(result);
+
+            // Set the access token as an environment variable
+            Environment.SetEnvironmentVariable(OAUTH_ACCESS_TOKEN_VAR_STRING, jObject["access_token"].ToString(), EnvironmentVariableTarget.User);
+
+            // Set the refresh token as an environment variable
+            Environment.SetEnvironmentVariable(OAUTH_REFRESH_TOKEN_VAR_STRING, jObject["refresh_token"].ToString(), EnvironmentVariableTarget.User);
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine("Error getting access token: " + ex.Message);
+            throw;
+        }
+    }
+
+    static string GetAuthHost()
+    {
+        return RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ? "LOCAL" : "HOST '0.0.0.0/0'";

```

Comment:
Should we allow for the sample app user to specify where their IDP is hosted?

Reply:
I think simplifying the 'host getting' is good here. I made a comment that these are default and can be changed.